src/librustc_trans/trans/adt.rs

```diff
@@ -373,10 +376,10 @@ fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
 
         // Perhaps one of the fields of this struct is non-zero
         // let's recurse and find out
-        ty::ty_struct(def_id, substs) => {
-            let fields = ty::lookup_struct_fields(tcx, def_id);
+        ty::ty_struct(def, substs) => {
+            let fields = &def.variants[0].fields[];
             for (j, field) in fields.iter().enumerate() {
-                let field_ty = ty::lookup_field_type(tcx, def_id, field.id, substs);
+                let field_ty = field.ty.get().subst(tcx, substs);
```

Comment:
Maybe `.ty.get()` would be better off as `.ty()` - can the `ty` field be fully private?
Would be nice to know one cannot simply `field.ty.set(ty)`.